listenbrainz/webserver/static/js/src/BrainzPlayer.test.tsx

```diff
@@ -0,0 +1,271 @@
+import * as React from "react";
+import { shallow, mount } from "enzyme";
+import BrainzPlayer, { DataSourceType } from "./BrainzPlayer";
+import YoutubePlayer from "./YoutubePlayer";
+import SpotifyPlayer from "./SpotifyPlayer";
+import APIService from "./APIService";
+
+const props = {
+  spotifyUser: {
+    access_token: "heyo",
+    permission: "read" as SpotifyPermission,
+  },
+  direction: "up" as BrainzPlayDirection,
+  onCurrentListenChange: (listen: Listen) => {},
+  listens: [],
+  newAlert: (
+    alertType: AlertType,
+    title: string,
+    message: string | JSX.Element
+  ) => {},
+  apiService: new APIService("base-uri"),
+};
+
+// Give yourself a two minute break and go listen to this gem
+// https://musicbrainz.org/recording/7fcaf5b3-e682-4ce6-be61-d3bce775a43f
+const listen: Listen = {
+  listened_at: 0,
+  track_metadata: {
+    artist_name: "Moondog",
+    track_name: "Bird's Lament",
+  },
+};
+
+describe("BrainzPlayer", () => {
+  it("renders correctly", () => {
+    const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
+    expect(wrapper.html()).toMatchSnapshot();
+  });
+
+  it("creates a Youtube datasource", () => {
+    const mockProps = {
+      ...props,
+      spotifyUser: {},
+    };
+    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
+    const instance = wrapper.instance();
+    expect(instance.dataSources).toHaveLength(1);
+    expect(instance.dataSources[0].current).toBeInstanceOf(YoutubePlayer);
+  });
+
+  it("creates a Spotify datasource when passed a spotify user with right permissions", () => {
+    const mockProps = {
+      ...props,
+      spotifyUser: {
+        access_token: "haveyouseenthefnords",
+        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
```

Comment:
the SpotifyPermission type only contains individual permissions like `streaming` and `user-read-email`. I think we should change the code so that `spotifyUser.permission` is  a list of `SpotifyPermission`.

Reply:
Yeah, I think that would be a good idea.
Maybe in another PR?
I can do the front-end if you want to do the back-end?

Reply:
Sounds great to me! Could you please open a ticket for this, so we don't forget?